### PR TITLE
[oneDPL] radix sort: workload optimization

### PR DESCRIPTION
**Pull request by MikeDvorskiy**
A macro _ONEDPL_RADIX_WORKLOAD_TUNING enables following workload tunning:

//To achieve better performance number of segments and block size are variated depend on data size (32bits):
//1. 32K...512K  - number of segments is increased up to 8 times
//2. 512K...2M   - number of segments is increased up up 4 times
//3. 2M...16M... - work-group size (count phase) is increased up to 128

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h [hunk 1]
```diff
@@ -25,6 +25,12 @@
 #include "parallel_backend_sycl_utils.h"
 #include "execution_sycl_defs.h"
 
+#define _ONEDPL_RADIX_WORKLOAD_TUNING 1
+//To achieve the better occupancy work group size and block size are variated depend on data size (32bits):
+//1. 32K...512K  - an estimated workgroup size is reduced down to 8 times
+//2. 512K...2M   - an estimated workgroup size is reduced down to 4 times
+//3. 2M...16M... - block size is increased up to 128
```
[the remark this answers is not in the document]
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:32]: The comment tells about "max work group size". In its turn is used to calculate number of segments. Other words, number of segments is increased.
`const ::std::size_t __segments = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __wg_size);`

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h [hunk 2]
```diff
@@ -610,6 +616,11 @@ struct __parallel_radix_sort_iteration
         // block size must be a power of 2 and not less than the number of states.
         // TODO: Check how to get rid of that restriction.
         __block_size = sycl::max(oneapi::dpl::__internal::__dpl_bit_floor(__block_size), ::std::size_t(__radix_states));
+#if _ONEDPL_RADIX_WORKLOAD_TUNING
+        const ::std::size_t __n = __in_rng.size();
+        if(__n > (1<<21)/*2M*/)
+            __block_size = 128;
```
[the remark this answers is not in the document]
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:624]: Looking into the code.. actually, we have a mess with names and their meaning...
For example in count phase block size is SYCL work group size:-)

Lets arranged the  definitions. If we talking about SYCL iteration space I suggest using the standard well-known terms: `work group count`, `work group size`, `sub group size`, `sub group count per work group` and N  - total number of elements passed to an algorithm to be processed.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h [hunk 3]
```diff
@@ -25,6 +25,12 @@
 #include "parallel_backend_sycl_utils.h"
 #include "execution_sycl_defs.h"
 
+#define _ONEDPL_RADIX_WORKLOAD_TUNING 1
+//To achieve better performance number of segments and block size are variated depend on data size (32bits):
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:29]: Could you update the comment? block size is not used anymore. Could you also clarify the meaning of "(32bits)"? 
Also, can you replace "depend on" to "depending on"?

I guess the performance impact was evaluated only for 32-bit types according to the comment. What do you think about measuring it for other types (e.g. 16-bit and 64-bit wide ones)? Probably, the constant (128) may be different for them.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h [hunk 4]
```diff
@@ -582,11 +589,15 @@ struct __parallel_radix_sort_iteration
...
         ::std::size_t __max_sg_size = oneapi::dpl::__internal::__max_sub_group_size(__exec);
-        ::std::size_t __scan_wg_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
-        ::std::size_t __block_size = __max_sg_size;
         ::std::size_t __reorder_sg_size = __max_sg_size;
+        ::std::size_t __scan_wg_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+#if _ONEDPL_RADIX_WORKLOAD_TUNING
+        ::std::size_t __count_wg_size = (__n > (1<<21)/*2M*/ ? 128 : __max_sg_size);
+#else
+        ::std::size_t __count_wg_size = __max_sg_size;
+#ednif
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:598]: a typo: ednif -> endif

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h (hunk 3 above)
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:29]: 32bits is more popular case I guess... But yes, it make sense to be estimated and probably to have another constant.

Comment (dmitriy-sobolev) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h]: I checked `count` kernel and found that the amount of allocated SLM does no depend on a input data type. So the performance difference for other input types is unlikely.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h [hunk 5]
```diff
@@ -582,11 +589,15 @@ struct __parallel_radix_sort_iteration
                                                                         __decay_t<_InRange>, __decay_t<_OutRange>>;
 
         ::std::size_t __max_sg_size = oneapi::dpl::__internal::__max_sub_group_size(__exec);
-        ::std::size_t __scan_wg_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
-        ::std::size_t __block_size = __max_sg_size;
         ::std::size_t __reorder_sg_size = __max_sg_size;
+        ::std::size_t __scan_wg_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+#if _ONEDPL_RADIX_WORKLOAD_TUNING
+        ::std::size_t __count_wg_size = (__n > (1<<21)/*2M*/ ? 128 : __max_sg_size);
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:595]:
```suggestion
        ::std::size_t __count_wg_size = (__in_rng.size() > (1<<21)/*2M*/ ? 128 : __max_sg_size);
```

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h (hunk 4 above)
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:598]:
```suggestion
#endif
```

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h [hunk 6]
```diff
@@ -25,6 +25,12 @@
 #include "parallel_backend_sycl_utils.h"
 #include "execution_sycl_defs.h"
 
+#define _ONEDPL_RADIX_WORKLOAD_TUNING 1
+//To achieve better performance number of segments and work-group size  are variated depending on number of elements:
+//1. 32K...512K  - number of segments is increased up to 8 times
+//2. 512K...2M   - number of segments is increased up up 4 times
+//3. 2M...16M... - work-group size (count phase) is increased up to 128
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:32]:
```suggestion
//To achieve better performance, number of segments and work-group size are variated depending on a number of elements:
//1. 32K...512K  - number of segments is increased up to 8 times
//2. 512K...2M   - number of segments is increased up up 4 times
//3. 2M... - work-group size (count phase) is increased up to 128
```

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h [hunk 7]
```diff
@@ -728,15 +740,17 @@ __parallel_radix_sort(_ExecutionPolicy&& __exec, _Range&& __in_rng, _Proj __proj
...
-        __event = __subgroup_radix_sort<_RadixSortKernel, __wg_size * 16, 32, __radix_bits, __is_ascending>{}(
-            __exec.queue(), ::std::forward<_Range>(__in_rng), __proj);
     else
     {
         constexpr ::std::uint32_t __radix_iters = __get_buckets_in_type<_KeyT>(__radix_bits);
         const ::std::uint32_t __radix_states = 1 << __radix_bits;
 
-        const ::std::size_t __wg_size = __max_wg_size;
+#if _ONEDPL_RADIX_WORKLOAD_TUNING
+        const auto __wg_sz_k = __n >= (1<<15)/*32K*/ && __n <  (1<<19)/*512K*/ ? 8 : __n <= (1<<21)/*2M*/ ? 4 : 1;
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:749]: Here is a suggestion according to clang-format:
```suggestion
        const auto __wg_sz_k =
            __n >= (1 << 15) /*32K*/ && __n < (1 << 19) /*512K*/ ? 8 : __n <= (1 << 21) /*2M*/ ? 4 : 1;
```

**Review by dmitriy-sobolev — APPROVED** [review]
LGTM

**Pull request merged**